### PR TITLE
Fix Cats Effect IO type inference

### PR DESCRIPTION
**Pull request by yangzai**
As mention in this [comment](https://github.com/sksamuel/elastic4s/pull/1981#issuecomment-702211419) the generalisation of implicits to provide `Executor`/`Functor` typeclass instances for all types with `Async`/`CatsFunctor` instances loses `IO` specificity and implicit precedence priority, making client execution infer to the default `Future` effect return type. This makes using the `IO` effect type very unwieldy, having to explicitly state all type parameters for client execution.

This PR brings back `IO` type inference without losing support for other `Async` effect type.

### CODE REVIEW

File: elastic4s-effect-cats-2/src/main/scala/com/sksamuel/elastic4s/cats/effect/instances/CatsEffectInstances.scala [hunk 1]
```diff
@@ -12,6 +12,9 @@ trait CatsEffectInstances {
 
   implicit def catsEffectExecutor[F[_]: Async]: Executor[F] =
     new CatsEffectExecutor[F]
+
+  //this needs to be at the bottom
```
Comment (yangzai) [inline elastic4s-effect-cats-2/src/main/scala/com/sksamuel/elastic4s/cats/effect/instances/CatsEffectInstances.scala:16]: I don't know why the ordering matters.

An alternative implementation with a lower priority trait `CatsEffectInstances0` also works:
```scala
trait CatsEffectInstances extends CatsEffectInstances0 {
  implicit def catsFunctor[F[_]: CatsFunctor]: Functor[F] = new Functor[F] {
    override def map[A, B](fa: F[A])(f: A => B): F[B] = CatsFunctor[F].map(fa)(f)
  }

  implicit def catsEffectExecutor[F[_]: Async]: Executor[F] =
    new CatsEffectExecutor[F]
}

trait CatsEffectInstances0 {
  implicit val ioExecutor: Executor[IO] = new CatsEffectExecutor[IO]
}
```
but I'm equally confused as to why the more specific `ioExecutor` needs to be in the lower and not the higher priority trait for it to work.

Comment (sksamuel) [reply elastic4s-effect-cats-2/src/main/scala/com/sksamuel/elastic4s/cats/effect/instances/CatsEffectInstances.scala:16]: I  want to do https://github.com/sksamuel/elastic4s/issues/2379 as well for 8.0. Maybe that's easier to do now ?

Comment (yangzai) [reply elastic4s-effect-cats-2/src/main/scala/com/sksamuel/elastic4s/cats/effect/instances/CatsEffectInstances.scala:16]: > I want to do #2379 as well for 8.0. Maybe that's easier to do now ?

I think they are mutually exclusive. For #2379 most of the work seem to revolve around refactoring `HttpClient`.

The main pain points with the current `execute` for me are:
- Defaulting to `Future`
The default `Future` typeclass instances in the companion objects for `Functor`/`Executor` seems conflict prone to me in terms of implicit resolution. Maybe you could drop the default instances and have 2 execution methods instead (`execute`/`executeF`) where one always return a `Future`. Or just keep to 1 and have users manually import `Future` instances.
2. The type parameters `[T, U, F[_]]`
Having to be explicit with the effect type `F` is fine but it forces you to have to figure out what `T` and `U` are. Being able to explicit on `F` without `T` and `U` would be nice, perhaps something in the direction of `client.effect[IO].execute(...)`

**Comment by sksamuel** [comment]
This is a little beyond my knowledge. What's the tl;dr of the change ?

File: elastic4s-effect-cats-2/src/test/scala/com.sksamuel.elastic4s.cats.effect/CatsEffectTest.scala [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import cats.effect.IO
+import com.sksamuel.elastic4s.{ElasticClient, ElasticProperties}
+import com.sksamuel.elastic4s.ElasticDsl._
+import org.scalatest.flatspec.AnyFlatSpec
+import com.sksamuel.elastic4s.cats.effect.instances._
+import com.sksamuel.elastic4s.http.JavaClient
+import com.sksamuel.elastic4s.requests.indexes.admin.DeleteIndexResponse
+import com.sksamuel.elastic4s.requests.indexes.{CreateIndexRequest, CreateIndexResponse, DeleteIndexRequest}
+
+class CatsEffectTest extends AnyFlatSpec {
```
Comment (yangzai) [inline elastic4s-effect-cats-2/src/test/scala/com.sksamuel.elastic4s.cats.effect/CatsEffectTest.scala:13]: @sksamuel basically, without the changes in `CatsEffectInstances` the first test case would not compile unless we rewrite it with explicit type params like the second test case.

Comment (sksamuel) [reply elastic4s-effect-cats-2/src/test/scala/com.sksamuel.elastic4s.cats.effect/CatsEffectTest.scala:13]: Do users of standard scala futures have any breaking changes ?

Comment (yangzai) [reply elastic4s-effect-cats-2/src/test/scala/com.sksamuel.elastic4s.cats.effect/CatsEffectTest.scala:13]: If users import `com.sksamuel.elastic4s.cats.effect.instances._` they intend `IO` to be the default effect. `Futures` should still work with explicit type params even with Cats Effect instances imported. I can add another test for that.

Comment (sksamuel) [reply elastic4s-effect-cats-2/src/test/scala/com.sksamuel.elastic4s.cats.effect/CatsEffectTest.scala:13]: Ok great. I just wanted to confirm that people who are not explicitly using effects would still be getting the default Futures implementation.

Comment (yangzai) [reply elastic4s-effect-cats-2/src/test/scala/com.sksamuel.elastic4s.cats.effect/CatsEffectTest.scala:16]: @sksamuel Test cases updated and branch rebased. I will need approval to trigger CI.

**Comment by sksamuel** [comment]
Triggered

**Pull request merged**